pkgs/dart_tooling_mcp_server/test/test_harness.dart

```diff
@@ -0,0 +1,283 @@
+///   between them. Includes a debug mode for running them both in process to
+///   allow for breakpoints, but the default mode is to run the server in a
+///   separate process.
+class TestHarness {
+  final TestProcess flutterProcess;
+  final DartToolingMCPClient mcpClient;
+  final ServerConnection mcpServerConnection;
+  final String dtdUri;
+
+  TestHarness._(
+    this.flutterProcess,
+    this.mcpClient,
+    this.mcpServerConnection,
+    this.dtdUri,
+  );
+
+  /// Starts a flutter process as well as an MCP client and server.
+  ///
+  /// Handles the initialization handshake between the MCP client and server as
+  /// well.
+  ///
+  /// By default this will run with the MCP server compiled as a separate binary
+  /// to mimic as closely as possible the real world behavior. This makes it so
+  /// breakpoints in the server don't work however, so you can set [debugMode]
+  /// to `true` and we will run it in process instead, which allows breakpoints
+  /// since everything is running in the same isolate.
+  static Future<TestHarness> start({
+    @Deprecated('For debugging only, do not submit') bool debugMode = false,
+  }) async {
+    final platform = Platform.isLinux
+        ? 'linux'
+        : Platform.isMacOS
+            ? 'macos'
+            : throw StateError(
+                'unsupported platform, only mac and linux are supported',
+              );
+    final flutterProcess = await TestProcess.start(
+      // TODO: Get flutter SDK location from somewhere.
+      'flutter',
+      ['run', '-d', platform],
+      workingDirectory: counterAppPath,
+    );
+    addTearDown(() async {
+      flutterProcess.stdin.writeln('q');
+      await flutterProcess.shouldExit(0);
+    });
+
+    final dtdUri = await _getDTDUri(flutterProcess);
+
+    final mcpClient = DartToolingMCPClient();
+    addTearDown(mcpClient.shutdown);
+    final connection = await _initializeMCPServer(mcpClient, debugMode);
+
+    final fakeEditorExtension = await FakeEditorExtension.connect(
+      flutterProcess,
+      dtdUri,
+    );
+    addTearDown(fakeEditorExtension.shutdown);
+
+    return TestHarness._(flutterProcess, mcpClient, connection, dtdUri);
+  }
+
+  /// Connects the MCP server to the dart tooling daemon at [dtdUri] using the
+  /// "connectDartToolingDaemon" tool function.
+  Future<void> connectToDtd() async {
+    final tools = (await mcpServerConnection.listTools()).tools;
+
+    final connectTool = tools.singleWhere(
+      (t) => t.name == 'connectDartToolingDaemon',
+    );
+
+    final result = await callToolWithRetry(
+      CallToolRequest(name: connectTool.name, arguments: {'uri': dtdUri}),
+    );
+    expect(result.isError, isNot(true), reason: result.content.join('\n'));
+  }
+
+  /// Sends [request] to [mcpServerConnection], retrying 5 times.
+  ///
+  /// Some methods will fail if the DTD connection is not yet ready.
+  Future<CallToolResult> callToolWithRetry(CallToolRequest request) async {
+    final maxTries = 10;
```

Comment:
should this be 5 or should the dart doc comment be changed to 10?

Reply:
Updated to 5 (had changed to 10 just when debugging), and I made this a parameter also.